src/Passes/Source/Rules/OperandPrototype.hpp

```diff
@@ -0,0 +1,87 @@
+#pragma once
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Llvm/Llvm.hpp"
+
+#include <unordered_map>
+#include <vector>
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    /// IOperandPrototype describes an IR pattern and allows matching against
+    /// LLVMs llvm::Value type.
+    class IOperandPrototype
+    {
+      public:
+        using Instruction = llvm::Instruction;
+        using String      = std::string;
+        using Value       = llvm::Value;
+        using Child       = std::shared_ptr<IOperandPrototype>;
+        using Children    = std::vector<Child>;
+        using Captures    = std::unordered_map<std::string, Value*>;
+
+        /// Constructors and desctructors
+        /// @{
+        IOperandPrototype() = default;
+        virtual ~IOperandPrototype();
+        /// @}
+
+        /// Interface functions
+        /// @{
+        virtual bool  match(Value* value, Captures& captures) const = 0;
+        virtual Child copy() const                                  = 0;
+        /// @}
+
+        /// Shared functionality
+        /// @{
+
+        /// Adds a child to be matched against the matchees children. Children
+        /// are matched in order and by size.
+        void addChild(Child const& child);
+
+        /// Flags that this operand should be captured. This function ensures
```

Comment:
The comment alone here is not sufficient to understand what the purpose is of this flag; what operands can and should be captured? What happens with captured operands? A see also reference to the piece of code that uses this information might help to clarify as well.

Reply:
I think what you possible need is some high level description of the `OperandPrototype`, its relation to creating rules and in that context. what captures are?

Reply:
Yes.